pulsar-client/src/main/java/org/apache/pulsar/client/impl/transaction/TransactionImpl.java

```diff
@@ -123,8 +132,25 @@ public CompletableFuture<Void> registerProducedTopic(String topic) {
         return completableFuture;
     }
 
-    public synchronized void registerSendOp(CompletableFuture<MessageId> sendFuture) {
-        sendFutureList.add(sendFuture);
+    public void registerSendOp(CompletableFuture<MessageId> newSendFuture) {
+        sendLock.lock();
+        try {
+            if (sendCount.getAndIncrement() == 0) {
+                sendFuture = new CompletableFuture<>();
+            }
+        } finally {
+            sendLock.unlock();
+        }
+        newSendFuture.thenRun(() -> {
+            sendLock.lock();
```

Comment:
why do we need this lock?

Reply:
1. decrease the opCount 1 -> 0 
2. check opCount == 0, and then increase the opCount 0 -> 1
3. future = new future;
4. future.complete
5. finally, the future is completely and opCount is 1